### PR TITLE
planner: fix DATA RACE on global binding

### LINKED ISSUE #58616: DATA RACE InitBindingHandle
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 2. What did you expect to see? (Required)

```
cd tests/integrationtest
./run-tests.sh -r planner/core/casetest/physicalplantest/physical_plan

```
### 3. What did you see instead (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c007b20ca8 by main goroutine:
  github.com/pingcap/tidb/pkg/domain.(*Domain).InitBindingHandle()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/domain.go:2089 +0x64
  github.com/pingcap/tidb/pkg/session.bootstrapSessionImpl()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:3526 +0x580
  github.com/pingcap/tidb/pkg/session.BootstrapSession()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:3408 +0x290
  main.createStoreDDLOwnerMgrAndDomain()
      /Users/weizhenwang/devel/opensource/tidb/cmd/tidb-server/main.go:419 +0x26c
  main.main()
      /Users/weizhenwang/devel/opensource/tidb/cmd/tidb-server/main.go:322 +0x7f0

Previous read at 0x00c007b20ca8 by goroutine 2496:
  github.com/pingcap/tidb/pkg/domain.(*Domain).BindHandle()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/domain.go:2083 +0x78
  github.com/pingcap/tidb/pkg/planner.init.0.func1()
      /Users/weizhenwang/devel/opensource/tidb/pkg/planner/optimize.go:594 +0x18
  github.com/pingcap/tidb/pkg/bindinfo.matchSQLBinding()
      /Users/weizhenwang/devel/opensource/tidb/pkg/bindinfo/binding.go:146 +0x2bc
  github.com/pingcap/tidb/pkg/bindinfo.MatchSQLBinding()
      /Users/weizhenwang/devel/opensource/tidb/pkg/bindinfo/binding.go:114 +0xdd8
  github.com/pingcap/tidb/pkg/planner.Optimize()
      /Users/weizhenwang/devel/opensource/tidb/pkg/planner/optimize.go:243 +0xdc8
  github.com/pingcap/tidb/pkg/executor.(*Compiler).Compile()
      /Users/weizhenwang/devel/opensource/tidb/pkg/executor/compiler.go:101 +0x594
  github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:2101 +0xa1c
  github.com/pingcap/tidb/pkg/session.(*session).ExecuteInternalStmt()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:2005 +0x128
  github.com/pingcap/tidb/pkg/session.(*session).ExecRestrictedSQL.func1()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:1970 +0x1f0
  github.com/pingcap/tidb/pkg/session.(*session).withRestrictedSQLExecutor()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:1956 +0x2d0
  github.com/pingcap/tidb/pkg/session.(*session).ExecRestrictedSQL()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:1960 +0xc4
  github.com/pingcap/tidb/pkg/resourcegroup/runaway.ExecRCRestrictedSQL()
      /Users/weizhenwang/devel/opensource/tidb/pkg/resourcegroup/runaway/record.go:382 +0x2e8
  github.com/pingcap/tidb/pkg/domain.(*RUStatsWriter).isLatestDataInserted()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/ru_stats.go:220 +0x184
  github.com/pingcap/tidb/pkg/domain.(*RUStatsWriter).DoWriteRUStatistics()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/ru_stats.go:142 +0xc4
  github.com/pingcap/tidb/pkg/domain.(*Domain).requestUnitsWriterLoop()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/ru_stats.go:81 +0x228
  github.com/pingcap/tidb/pkg/domain.(*Domain).requestUnitsWriterLoop-fm()
      <autogenerated>:1 +0x34
  github.com/pingcap/tidb/pkg/util.(*WaitGroupEnhancedWrapper).Run.func1()
      /Users/weizhenwang/devel/opensource/tidb/pkg/util/wait_group_wrapper.go:103 +0xa8

Goroutine 2496 (running) created at:
  github.com/pingcap/tidb/pkg/util.(*WaitGroupEnhancedWrapper).Run()
      /Users/weizhenwang/devel/opensource/tidb/pkg/util/wait_group_wrapper.go:98 +0x12c
  github.com/pingcap/tidb/pkg/domain.(*Domain).Start()
      /Users/weizhenwang/devel/opensource/tidb/pkg/domain/domain.go:1541 +0x5bc
  github.com/pingcap/tidb/pkg/session.bootstrapSessionImpl()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:3515 +0x558
  github.com/pingcap/tidb/pkg/session.BootstrapSession()
      /Users/weizhenwang/devel/opensource/tidb/pkg/session/session.go:3408 +0x290
  main.createStoreDDLOwnerMgrAndDomain()
      /Users/weizhenwang/devel/opensource/tidb/cmd/tidb-server/main.go:419 +0x26c
  main.main()
      /Users/weizhenwang/devel/opensource/tidb/cmd/tidb-server/main.go:322 +0x7f0
==================

```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #58616

Problem Summary: planner: fix DATA RACE on global binding

### What changed and how does it work?

planner: fix DATA RACE on global binding

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/domain.go [hunk 1]
```diff
@@ -146,7 +146,7 @@ type Domain struct {
 	store           kv.Storage
 	infoCache       *infoschema.InfoCache
 	privHandle      *privileges.Handle
-	bindHandle      bindinfo.GlobalBindingHandle
+	bindHandle      atomic.Value
```
Comment (hawkingrei) [inline pkg/domain/domain.go:149]:
```suggestion
	bindHandle      atomic.Pointer[bindinfo.GlobalBindingHandle]
```

Comment (hawkingrei) [reply pkg/domain/domain.go:149]: This avoids type conversion

Comment (qw4990) [reply pkg/domain/domain.go:149]: This is an interface, not a structure, so we can't use "Pointer" here.

**Review by hawkingrei — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Pull request merged**